module/zfs/metaslab.c

```diff
@@ -2598,8 +2881,145 @@ metaslab_group_alloc_verify(spa_t *spa, const blkptr_t *bp, void *tag)
 #endif
 }
 
+/*
+ * block_alloc_stats() is common to metaslab layer and zdb (for auditing).
+ */
```

Comment:
maybe note that the `ms_alloc_stats_t` is (indirectly) stored on disk, so this algorithm is part of the on-disk format.

Reply:
Good point. Updated the comments